pageserver/src/storage_sync.rs

```diff
@@ -129,1441 +114,661 @@
-            )
-        })?
-        .with_context(|| {
-            format!(
-                "Failed to write remote metadata bytes locally to path '{}'",
-                local_metadata_path.display()
-            )
-        })?;
-    } else {
-        info!("Local metadata at path '{}' has later disk consistent Lsn ({local_lsn:?}) than the remote one ({remote_lsn}), skipping the update", local_metadata_path.display());
+
+        // Launch the tasks immediately, if possible
+        self.launch_queued_tasks(upload_queue);
+        Ok(())
     }
 
-    Ok(())
-}
+    ///
+    /// Wait for all previously scheduled uploads/deletions to complete
+    ///
+    pub async fn wait_completion(self: &Arc<Self>) -> anyhow::Result<()> {
+        let (sender, mut receiver) = tokio::sync::watch::channel(());
+        let barrier_op = UploadOp::Barrier(sender);
 
-async fn delete_timeline_data(
-    conf: &'static PageServerConf,
-    (storage, index, sync_queue): (&GenericRemoteStorage, &RemoteIndex, &SyncQueue),
-    sync_id: TenantTimelineId,
-    mut new_delete_data: SyncData<LayersDeletion>,
-    sync_start: Instant,
-) {
-    static TASK_NAME: &str = "delete";
-
-    let timeline_delete = &mut new_delete_data.data;
-
-    if !timeline_delete.deletion_registered {
-        if let Err(e) = update_remote_data(
-            conf,
-            storage,
-            index,
-            sync_id,
-            RemoteDataUpdate::Delete(&timeline_delete.layers_to_delete),
-        )
-        .await
         {
-            error!("Failed to update remote timeline {sync_id}: {e:?}");
-            new_delete_data.retries += 1;
-            sync_queue.push(sync_id, SyncTask::Delete(new_delete_data));
-            register_sync_status(sync_id, sync_start, TASK_NAME, Some(false));
-            return;
+            let mut upload_queue = self.upload_queue.lock().unwrap();
+            ensure!(
+                upload_queue.latest_metadata.is_some(),
+                "upload queue not initialized"
+            );
+
+            upload_queue.queued_operations.push_back(barrier_op);
+
+            // Launch the task immediately, if possible
+            self.launch_queued_tasks(upload_queue);
         }
+
+        receiver.changed().await?;
+        Ok(())
     }
-    timeline_delete.deletion_registered = true;
 
-    let sync_status = delete_timeline_layers(storage, sync_queue, sync_id, new_delete_data).await;
-    register_sync_status(sync_id, sync_start, TASK_NAME, Some(sync_status));
-}
+    ///
+    /// Pick next tasks from the queue, and start as many of them as possible without violating
+    /// the ordering constraints.
+    ///
+    /// The caller needs to already hold the `upload_queue` lock.
+    fn launch_queued_tasks(self: &Arc<Self>, mut upload_queue: MutexGuard<UploadQueue>) {
+        while let Some(next_op) = upload_queue.queued_operations.front() {
+            // Can we run this task now?
+            let can_run_now = match next_op {
```

Comment:
It is crucial for liveness that the conditions checked in `can_run_now { ...}` and the corresponding code in `fn perform_upload_task` is in sync with each other.
Basically, the programmer has to make sure that `can_run_now` is a pure function that is bound to return true.
Relatively brittle in my opinion.

I think a wait-for graph would be more robust.

But to get this merged quickly, we'll stick with what we have here.

Maybe we can throw in a watchdog to produce `warn!("upload queue clogged")` messages if the queue doesn't make forward progress.
Forward progress could be defined like "is it still the same element in `queued_operations.front()`?"

Reply:
If `inprogress_tasks` is empty, we can always proceed. That's implicit now, each branch in the `match` statement will return true if `inprogress_tasks` is empty, but perhaps it would be more clear to check for that before the other conditions. Something like:
```
    let can_run_now = if upload_queue.inprogress_tasks.is_empty() {
        // If the upload queue is empty, we can always start a new task
        true
    } else {
        // Can the next task in the queue run before the in-progress tasks have finished? 
        match next_op {
            ...
        }
    };
```

Reply:
Reconsidering this: all we should have before we ship this is a Prometheus metric for `inprogress_tasks.len()`, `queued_operations.len()`, etc.
That'll be sufficient to narrow down the problem if we overlooked something.